### PR TITLE
Live updates for object explorer

### PR DESCRIPTION
**Pull request by romainfrancois**
### Intent

part of #1994 

### Approach

Make sure a `View(x)` when called a second time updates the view, and removes the entry from `.rs.explorer.cache`. 

### Automated Tests

> Indicate whether this change includes unit tests or integration tests, or link a work item or pull request to add those tests to another repo. If the change cannot or will not be covered by a test, indicate why.

### QA Notes

In a new session, type: 

```r
f <- function() {
  x <- list(a = runif(1))
  View(x)
}
ls.str(.rs.explorer.cache)
f()
```

`x` is viewed with some number: 

image

`.rs.explorer.cache` is empty before the `f()` call: 

image

Then it has one entry after the call to `f()`: 

image

Call `f()` a second time, the view is replaced: 

image

`.rs.explorer.cache` still has only one entry, a new one: 

image

Without this change, `x` was not updated in the view, and entries were accumulating in `.rs.explorer.cache`, probably retaining the evaluation environment of `f()` forever

image

### Checklist

- [ ] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [ ] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [ ] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [ ] This PR passes all local unit tests

### CODE REVIEW

**Comment by romainfrancois** [comment]
draft because `onDetectChanges()` needs some more work for: https://github.com/rstudio/rstudio/issues/1994#issuecomment-1209433655

**Comment by romainfrancois** [comment]
This adds a `onDetectChanges()` that checks if the object being `View()`ed has changed in its environment

```r
x <- list(a = 1)
View(x)
```

shows: 

image

..and then 

```r
x <- list(b = 2)
```

updates the view: 

image

This is modeled on what is done for data frames (`DataViewer.cpp`), probably the code in `.rs.explorer.refresh()` could be inlined in the C++ ...

File: src/cpp/session/modules/SessionObjectExplorer.cpp [hunk 1]
```diff
@@ -193,6 +194,80 @@ void onDeferredInit(bool)
...
+   if (!core::thread::isMainThread())
+      return;
+
+   // unlikely that data will change outside of a REPL
+   if (source != module_context::ChangeSourceREPL) 
+      return;
+
+   Error error;
+   r::sexp::Protect rProtect;
+   SEXP envCache;
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionObjectExplorer.cpp:210]: Nit: initialize this to `R_NilValue` just to ensure it always has a valid R value.

File: src/cpp/session/modules/SessionObjectExplorer.cpp [hunk 2]
```diff
@@ -193,6 +194,80 @@ void onDeferredInit(bool)
...
+   if (error)
+   {
+      LOG_ERROR(error);
+      return;
+   }
+   
+   for (const std::string& id: cached)
+   {
+      SEXP s = Rf_install(id.c_str());
+      SEXP entry = Rf_findVarInFrame(envCache, s);
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionObjectExplorer.cpp:229]: Can we add a safety check here (make sure it's a list of length 5?)

**Review by kevinushey — COMMENTED** [review]
LGTM! I'd also recommend adding a couple sanity checks in the `onDetectChanges()` implementation.

**Pull request merged**